mobile/yggmail/yggmail.go

```diff
@@ -23,67 +24,106 @@ import (
 
 type peerAddrList []string
 
-const ERROR_OPEN_DB int = 1
-const ERROR_PASSWORD int = 2
-const ERROR_START int = 3
-const ERROR_SMTP int = 4
-const ERROR_OVERLAY_SMTP int = 5
-const ERROR_IMAP int = 6
-const ERROR_AUTH int = 7
+// constants defining error types
+const (
+	ERROR_OPEN_DB = iota
+	ERROR_PASSWORD
+	ERROR_START
+	ERROR_SMTP
+	ERROR_OVERLAY_SMTP
+	ERROR_IMAP
+	ERROR_AUTH
+)
+
+// constants defining Yggmail.state
+const (
+	Running = iota
+	ShuttingDown
+	Stopped
+	Error
+)
 
 type Yggmail struct {
 	storage           *sqlite3.SQLite3Storage
 	imapServer        *imapserver.IMAPServer
 	localSmtpServer   *smtp.Server
 	overlaySmtpServer *smtp.Server
+	transport         *transport.YggdrasilTransport
 	DatabaseName      string
 	Logger            Logger
 	AccountName       string
+	state             int
+	locker            sync.Mutex
 }
 
-func (ym *Yggmail) OpenDatabase() {
-	if err := ym.openDatabase(); err != nil {
-		ym.sendError(ERROR_OPEN_DB, "%s", *err)
-	}
+func (ym *Yggmail) GetState() int {
+	ym.locker.Lock()
+	defer ym.locker.Unlock()
+	return ym.state
 }
 
-func (ym *Yggmail) openDatabase() *error {
-	storage, err := sqlite3.NewSQLite3StorageStorage(ym.DatabaseName)
-	if err != nil {
-		return &err
+func (ym *Yggmail) setState(stateConst int) {
+	if stateConst == Running || stateConst == ShuttingDown || stateConst == Stopped || stateConst == Error {
+		ym.locker.Lock()
+		ym.state = stateConst
+		ym.locker.Unlock()
```

Comment:
I think you can use `defer` here as in the `GetState()` function. There will be almost no difference, but the code will look consistent.